javalib/src/main/scala/java/net/HttpCookie.scala

```diff
@@ -0,0 +1,374 @@
+    while (i < name.length()) {
+      val char = name.charAt(i)
+      if (char < 32 || // CTL
+          char == 127 || // CTL
+          char == '(' ||
+          char == ')' ||
+          char == '<' ||
+          char == '>' ||
+          char == '@' ||
+          char == ',' ||
+          char == ';' ||
+          char == ':' ||
+          char == '\\' ||
+          char == '"' ||
+          char == '/' ||
+          char == '[' ||
+          char == ']' ||
+          char == '?' ||
+          char == '=' ||
+          char == '{' ||
+          char == '}' ||
+          char == ' ' ||
+          char == '\t') {
+        throw new IllegalArgumentException("Illegal cookie name")
+      }
+
+      i += 1
+    }
+
+    name
+
+  }
+
+}
+
+final class HttpCookie private (
+    private var _comment: String = null,
+    private var _commentURL: String = null,
+    private var _httpOnly: Boolean = false,
+    private var _discard: Boolean = false,
+    private var _domain: String = null,
+    private var _maxAge: Long = -1,
+    private var _name: String = null,
+    private var _path: String = null,
+    private var _portlist: String = null,
+    private var _secure: Boolean = false,
+    private var _value: String = null,
+    private var _version: Int = 1
+) extends Cloneable {
+
+  def this(name: String, value: String) = {
+    this(
+      _name =
+        (if (name.length() == 0) {
+           throw new IllegalArgumentException("Illegal cookie name")
+         } else
+           HttpCookie.validateName(name)),
+      _value = value
+    )
+  }
+
+  override def equals(obj: Any): Boolean = {
+
+    if (obj.isInstanceOf[HttpCookie]) {
+      val other = obj.asInstanceOf[HttpCookie]
+      this.getComment().equals(other.getComment()) &&
+        this.getCommentURL().equals(other.getCommentURL()) &&
+        this.isHttpOnly().equals(other.isHttpOnly()) &&
+        this.getDiscard().equals(other.getDiscard()) &&
+        this.getDomain().equals(other.getDomain()) &&
+        this.getMaxAge().equals(other.getMaxAge()) &&
+        this.getName().equals(other.getName()) &&
+        this.getPath().equals(other.getPath()) &&
+        this.getPortlist().equals(other.getPortlist()) &&
+        this.getSecure().equals(other.getSecure()) &&
+        this.getValue().equals(other.getValue()) &&
+        this.getVersion().equals(other.getVersion())
+    } else {
+      false
+    }
+
+  }
+
+  def getComment(): String = this._comment
+  def getCommentURL(): String = this._commentURL
+  def getDiscard(): Boolean = this._discard
+  def getDomain(): String = this._domain
+  def getMaxAge(): Long = this._maxAge
+  def getName(): String = this._name
+  def getPath(): String = this._path
+  def getPortlist(): String = this._portlist
+  def getSecure(): Boolean = this._secure
+  def getValue(): String = this._value
+  def getVersion(): Int = this._version
+  def hasExpired(): Boolean = this._maxAge == 0
+  override def hashCode(): Int = toString().hashCode()
```

Comment:
This works and balances `equals`.

There is a recommendation that I have never quite understood as to
using `.##` instead of `hashCode` when defining a hashCode method.
I believe that `String` (and most other objects) have such.  I believe
that `.##` handles some special cases.

Reply:
There a are a number of `override hashCode` implementations
in the javalib code base, so it is hard to point out one which
is 'Best (or at least current) Practice'. Answering the question of
'why is that one better (or gives a better distribution)' could take
a semester course in Maths.

The one in `java.util.Arrays` gives a basic structure.  It comes from
Scala.js, where folks are knowledgeable, experienced, and super strict.

A non array body would look something like the 
example from `nio/file/attribute/PosixFileAttributeViewImpl.scala` 
```
   override def hashCode(): Int = {
      var res = 1
      res = 31 * res + deviceId.##
      res = 31 * res + inodeNumber.##
      res
    }
```
The idea is a field-by-field comparison by accumulating and shifting the intermediate results.  If the fields can be null, one must handle that (long discussion).

If I live a Good Life, Someday I may be saved from having to write 
equals/hashcode pairs.

Reply:
The JVM seems to include the name, value, domain and path into the hashcode, so I changed mine to do so as well, using the formula you gave there.